schema/loader.go

```diff
@@ -0,0 +1,165 @@
+package schema
+
+import (
+	"bytes"
+	"embed"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/santhosh-tekuri/jsonschema/v5"
+	"github.com/santhosh-tekuri/jsonschema/v5/httploader"
+)
+
+var (
+	//go:embed known_schemas
+	knownSchemas embed.FS
+)
+
+type (
+	SchemaFile string
+)
+
+const (
+	schemaDirectory = "known_schemas/"
+
+	// Presentation Exchange Schemas
+	PresentationDefinitionSchema              SchemaFile = "pe-presentation-definition.json"
+	PresentationDefinitionEnvelopeSchema      SchemaFile = "pe-presentation-definition-envelope.json"
+	PresentationSubmissionSchema              SchemaFile = "pe-presentation-submission.json"
+	PresentationClaimFormatDesignationsSchema SchemaFile = "pe-definition-claim-format-designations.json"
+	SubmissionClaimFormatDesignationsSchema   SchemaFile = "pe-submission-claim-format-designations.json"
+	SubmissionRequirementSchema               SchemaFile = "pe-submission-requirement.json"
+	SubmissionRequirementsSchema              SchemaFile = "pe-submission-requirements.json"
+
+	PresentationClaimFormatDesignationFormatDefinition SchemaFile = "pe-submission-claim-format-designations.json"
+
+	// Credential Manifest Schemas
+	CredentialManifestSchema    SchemaFile = "cm-credential-manifest.json"
+	CredentialApplicationSchema SchemaFile = "cm-credential-application.json"
+	CredentialResponseSchema    SchemaFile = "cm-credential-response.json"
+	OutputDescriptorsSchema     SchemaFile = "cm-output-descriptors.json"
+
+	// Wallet Rendering Schemas
+	DisplayMappingObjectSchema        SchemaFile = "wr-display-mapping-object.json"
+	EntityStylesSchema                SchemaFile = "wr-entity-styles.json"
+	LabeledDisplayMappingObjectSchema SchemaFile = "wr-labeled-display-mapping-object.json"
+
+	// VC JSON Schema Schemas
+	VerifiableCredentialJSONSchemaSchema SchemaFile = "vc-json-schema.json"
+)
+
+func (s SchemaFile) String() string {
+	return string(s)
+}
+
+// CachingLoader is a struct that holds local schemas
+type CachingLoader struct {
+	schemas map[string]string
+}
+
+// NewCachingLoader returns a new CachingLoader that enables the ability to cache http and https schemas
+func NewCachingLoader(schemas map[string]string) (*CachingLoader, error) {
+	cl := CachingLoader{schemas: make(map[string]string)}
+	for schemaURI, schema := range schemas {
+		if _, ok := cl.schemas[schemaURI]; ok {
+			return nil, fmt.Errorf("schema %q already exists", schemaURI)
+		}
+		cl.schemas[schemaURI] = schema
+	}
+	return &cl, nil
+}
+
+// EnableHTTPCache enables caching of http and https schemas
+func (cl *CachingLoader) EnableHTTPCache() {
+	jsonschema.Loaders["http"] = cl.cachingLoaderForProtocol("http")
+	jsonschema.Loaders["https"] = cl.cachingLoaderForProtocol("https")
+}
+
+func (cl *CachingLoader) cachingLoaderForProtocol(protocol string) func(url string) (io.ReadCloser, error) {
+	var m sync.Mutex
```

Comment:
Did you consider using a `sync.RWMutex` in order to lock more granularly? It could increase read throughput (though that's a small optimization).